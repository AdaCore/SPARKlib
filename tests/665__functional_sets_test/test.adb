with Ada.Text_IO;
with SPARK.Big_Integers;
with Inst; use Inst.Int_Sets;
with SPARK.Containers.Functional.Sets;

procedure Test with SPARK_Mode is

   procedure Assert (B : Boolean; S : String) with
     Pre => B;
   procedure Assert (B : Boolean; S : String) is
   begin
      if not B then
         Ada.Text_IO.Put_Line (S);
      end if;
   end Assert;

   procedure Test_Contains is
      X : Set;
   begin
      Assert (not Contains (X, 5), "Contains on empty sequence");
      for I in 1 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (Contains (X, 5) and Contains (X, 105), "Contains on element of the set");
      Assert (not Contains (X, 9) and not Contains (X, 109), "Contains on element not in the set");
   end Test_Contains;

   procedure Test_Choose is
      X : Set;
   begin
      for I in 1 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (Contains (X, Choose (X)), "Choose returns an element in the set");
   end Test_Choose;

   procedure Test_Length is
      use SPARK.Big_Integers;
      X : Set;
   begin
      Assert (Length (X) = 0, "Length of empty set");
      for I in 1 .. 10 loop
         X := Add (X, I);
      end loop;
      Assert (Length (X) = 10, "Length of non-empty set");
   end Test_Length;

   procedure Test_Le is
      X, Y : Set;
   begin
      Assert (Y <= X, """<="" on empty sets");
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (Y <= X and not (X <= Y), """<="" on sets with different lengths");
      Y := X;
      Assert (Y <= X and X <= Y, """<="" on copied sets");
      X := Add (X, 1);
      Y := Add (Y, 101);
      Assert (Y <= X and X <= Y, """<="" on sets uses element equivalence");
      X := Add (X, 2);
      Assert (Y <= X and not (X <= Y), """<="" on smaller sets");
      Y := Add (Y, 103);
      Assert (not (Y <= X) and not (X <= Y), """<="" on different sets");
   end Test_Le;

   procedure Test_Eq is
      X, Y : Set;
   begin
      Assert (X = Y, """="" on empty sets");
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (X /= Y, """="" on sets with different lengths");
      Y := X;
      Assert (X = Y, """="" on copied sets");
      X := Add (X, 1);
      Y := Add (Y, 101);
      Assert (X = Y, """="" on sets uses element equivalence");
      X := Add (X, 2);
      Y := Add (Y, 103);
      Assert (X /= Y, """="" on different sets");
   end Test_Eq;

   procedure Test_Is_Empty is
      X : Set;
   begin
      Assert (Is_Empty (X), "Is_Empty on empty set");
      X := Add (X, 1);
      Assert (not Is_Empty (X), "Is_Empty on non-empty set");
   end Test_Is_Empty;

   procedure Test_Included_Except is
      X, Y : Set;
   begin
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Y := X;
      X := Add (X, 102);
      Y := Add (Y, 103);
      Assert (Included_Except (X, Y, 2), "Included_Except returns True");
      Assert (not Included_Except (X, Y, 5), "Included_Except returns False");
   end Test_Included_Except;

   procedure Test_Includes_Intersection is
      X, Y, Z : Set;
   begin
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Y := X;
      Z := X;
      X := Add (X, 102);
      Y := Add (Y, 103);
      Assert (Includes_Intersection (Z, X, Y), "Includes_Intersection returns True");
      Z := Remove (Z, 5);
      Assert (not Includes_Intersection (Z, X, Y), "Includes_Intersection returns False");
   end Test_Includes_Intersection;

   procedure Test_Included_In_Union is
      X, Y, Z : Set;
   begin
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      for I in 2 .. 5 loop
         Y := Add (Y, I);
      end loop;
      for I in 3 .. 6 loop
         Z := Add (Z, I);
      end loop;
      Assert (Included_In_Union (Z, X, Y), "Included_In_Union returns True");
      Z := Add (Z, 9);
      Assert (not Included_In_Union (Z, X, Y), "Included_In_Union returns False");
   end Test_Included_In_Union;

   procedure Test_Is_Singleton is
      X : Set;
   begin
      Assert (not Is_Singleton (X, 2), "Is_Singleton on empty set");
      X := Add (X, 102);
      Assert (Is_Singleton (X, 2), "Is_Singleton on singleton");
      X := Add (X, 103);
      Assert (not Is_Singleton (X, 3), "Is_Singleton on set with more than 1 element");
   end Test_Is_Singleton;

   procedure Test_Not_In_Both is
      X, Y, Z : Set;
   begin
      for I in 4 .. 6 loop
         X := Add (X, I);
      end loop;
      for I in 1 .. 3 loop
         Y := Add (Y, I);
      end loop;
      for I in 2 .. 5 loop
         Z := Add (Z, I);
      end loop;
      Assert (Not_In_Both (Z, X, Y), "Not_In_Both no overlap");
      X := Add (X, 101);
      Assert (Not_In_Both (Z, X, Y), "Not_In_Both no overlap in Container");
      X := Add (X, 102);
      Assert (not Not_In_Both (Z, X, Y), "Not_In_Both overlap");
   end Test_Not_In_Both;

   procedure Test_No_Overlap is
      X, Y : Set;
   begin
      for I in 4 .. 6 loop
         X := Add (X, I);
      end loop;
      for I in 1 .. 3 loop
         Y := Add (Y, I);
      end loop;
      Assert (No_Overlap (X, Y), "No_Overlap returns True");
      X := Add (X, 101);
      Assert (not No_Overlap (X, Y), "No_Overlap returns False");
   end Test_No_Overlap;

   procedure Test_Num_Overlaps is
      use SPARK.Big_Integers;
      X, Y : Set;
   begin
      Assert (Num_Overlaps (X, Y) = 0, "Num_Overlap on empty sets");
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (Num_Overlaps (X, Y) = 0 and Num_Overlaps (Y, X) = 0, "Num_Overlap on empty and non-empty sets");
      for I in 4 .. 5 loop
         Y := Add (Y, I);
      end loop;
      Assert (Num_Overlaps (X, Y) = 2 and Num_Overlaps (Y, X) = 2, "Num_Overlap on included sets");
      for I in 2 .. 3 loop
         Y := Add (Y, I);
      end loop;
      Assert (Num_Overlaps (X, Y) = 2 and Num_Overlaps (Y, X) = 2, "Num_Overlap on non-included sets");
      for I in 4 .. 5 loop
         Y := Remove (Y, I);
      end loop;
      Assert (Num_Overlaps (X, Y) = 0 and Num_Overlaps (Y, X) = 0, "Num_Overlap on disjoint sets");
   end Test_Num_Overlaps;

   procedure Test_Empty_Set is
      use SPARK.Big_Integers;
      X : Set := Empty_Set;
   begin
      Assert (Length (X) = 0 and not Contains (X, 1), "Empty_Set is empty");
   end Test_Empty_Set;

   procedure Test_Add is
      use SPARK.Big_Integers;
      X, Y : Set;

   begin
      X := Add (X, 1);
      X := Add (X, 2);
      Assert (Length (X) = 2, "Add increments length");
      Assert (Contains (X, 102), "Add new element is in the set");
      Assert (Contains (X, 101), "Add previous elements are preserved");
   end Test_Add;

   procedure Test_Remove is
      use SPARK.Big_Integers;
      X : Set;
   begin
      for I in 1 .. 10 loop
         X := Add (X, I);
      end loop;
      X := Remove (X, 5);
      Assert (Length (X) = 9, "Remove decrements length");
      Assert (not Contains (X, 105), "Remove removes the equivalence class");
      Assert (Contains (X, 103) and not Contains (X, 111), "Remove reserves other values");
   end Test_Remove;

   procedure Test_Intersection is
      X, Y, Z : Set;
   begin
      Assert (Is_Empty (Intersection (X, Y)), "Intersection on empty sets");
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (Is_Empty (Intersection (X, Y)) and Is_Empty (Intersection (Y, X)), "Intersection on empty and non-empty sets");
      for I in 4 .. 5 loop
         Y := Add (Y, I);
      end loop;
      Assert (Intersection (X, Y) = Y and Intersection (Y, X) = Y, "Intersection on included sets");
      Z := Y;
      for I in 2 .. 3 loop
         Y := Add (Y, I);
      end loop;
      Assert (Intersection (X, Y) = Z and Intersection (Y, X) = Z, "Intersection on non-included sets");
      for I in 4 .. 5 loop
         Y := Remove (Y, I);
      end loop;
      Assert (Is_Empty (Intersection (X, Y)) and Is_Empty (Intersection (Y, X)), "Intersection on disjoint sets");
   end Test_Intersection;

   procedure Test_Union is
      X, Y, Z : Set;
   begin
      Assert (Is_Empty (Union (X, Y)), "Union on empty sets");
      for I in 4 .. 7 loop
         X := Add (X, I);
      end loop;
      Assert (Union (X, Y) = X and Union (Y, X) = X, "Union on empty and non-empty sets");
      for I in 4 .. 5 loop
         Y := Add (Y, I);
      end loop;
      Assert (Union (X, Y) = X and Union (Y, X) = X, "Union on included sets");
      Z := X;
      for I in 2 .. 3 loop
         Y := Add (Y, I);
         Z := Add (Z, I);
      end loop;
      Assert (Union (X, Y) = Z and Union (Y, X) = Z, "Union on non-included sets");
      for I in 4 .. 5 loop
         Y := Remove (Y, I);
      end loop;
      Assert (Union (X, Y) = Z and Union (Y, X) = Z, "Union on disjoint sets");
   end Test_Union;

   procedure Test_Iteration is
      B : Boolean;
      X : Set;
   begin
      for I in 1 .. 10 loop
         X := Add (X, I);
      end loop;
      Assert (Get_Set (Iterate (X)) = X, "Get_Set on iterator");
      for E of Iterate (X) loop
         Assert (Contains (X, E), "iteration only covers elements of the set");
      end loop;
      B := False;
      for S in Iterate (X) loop
         pragma Loop_Invariant (B = not Contains (S, 5));
         if Inst.Equivalent (Choose (S), 5) then
            B := True;
         end if;
      end loop;
      Assert (B, "iteration covers all elements of the set");
   end Test_Iteration;

   procedure Test_Aggregate is
      use SPARK.Big_Integers;
      X : Set := [1, 2, 3, 4, 5];
   begin
      Assert (Length (X) = 5, "Aggregate has expected length");
      Assert (Contains (X, 1) and Contains (X, 105) and not Contains (X, 7), "Aggregate has expected elements");
   end Test_Aggregate;

begin
   Test_Contains;
   Test_Choose;
   Test_Length;
   Test_Le;
   Test_Eq;
   Test_Is_Empty;
   Test_Included_Except;
   Test_Includes_Intersection;
   Test_Included_In_Union;
   Test_Is_Singleton;
   Test_Not_In_Both;
   Test_No_Overlap;
   Test_Num_Overlaps;
   Test_Empty_Set;
   Test_Add;
   Test_Remove;
   Test_Intersection;
   Test_Union;
   Test_Iteration;
   Test_Aggregate;
end Test;
