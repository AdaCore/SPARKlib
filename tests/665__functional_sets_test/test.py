from subprocess import call
from test_support import no_crash
import os

contains_manual_proof = False
os.environ["SPARKLIB_BODY_MODE"] = "On"


if __name__ == "__main__":
    no_crash(sparklib=True, opt=["--no-inlining", "-P", "test.gpr"])

    call(["gprbuild", "-q", "-P", "test.gpr"])
    call(["./obj/test"])
