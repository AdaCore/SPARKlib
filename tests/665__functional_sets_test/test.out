inst.ads:3:13: info: implicit aspect Always_Terminates on "Equivalent" has been proved, subprogram will terminate
inst.ads:3:62: info: division check proved (CVC5: 1 VC)
inst.ads:3:74: info: division check proved (CVC5: 1 VC)
test.adb:18:07: info: initialization of "X" proved
test.adb:20:07: info: precondition proved (CVC5: 1 VC)
test.adb:21:23: info: unrolling loop
test.adb:22:15: medium: precondition might fail
test.adb:24:07: info: precondition proved (CVC5: 1 VC)
test.adb:25:07: medium: precondition might fail
test.adb:29:07: info: initialization of "X" proved
test.adb:31:23: info: unrolling loop
test.adb:32:15: medium: precondition might fail
test.adb:34:07: info: precondition proved (CVC5: 1 VC)
test.adb:34:28: info: precondition proved (CVC5: 1 VC)
test.adb:39:07: info: initialization of "X" proved
test.adb:41:07: info: precondition proved (CVC5: 1 VC)
test.adb:41:15: info: predicate check proved (Trivial: 1 VC)
test.adb:41:28: info: predicate check proved (Trivial: 1 VC)
test.adb:42:24: info: unrolling loop
test.adb:43:15: medium: precondition might fail
test.adb:45:07: info: precondition proved (CVC5: 1 VC)
test.adb:45:15: info: predicate check proved (Trivial: 1 VC)
test.adb:45:28: info: predicate check proved (Trivial: 1 VC)
test.adb:49:07: info: initialization of "X" proved
test.adb:49:10: info: initialization of "Y" proved
test.adb:51:07: info: precondition proved (CVC5: 1 VC)
test.adb:52:23: info: unrolling loop
test.adb:53:15: info: precondition proved (CVC5: 4 VC)
test.adb:55:07: info: precondition proved (CVC5: 1 VC)
test.adb:57:07: info: precondition proved (CVC5: 1 VC)
test.adb:58:12: info: precondition proved (CVC5: 1 VC)
test.adb:59:12: medium: precondition might fail
test.adb:60:07: info: precondition proved (Z3: 1 VC)
test.adb:61:12: info: precondition proved (Z3: 1 VC)
test.adb:62:07: medium: precondition might fail
test.adb:63:12: medium: precondition might fail
test.adb:64:07: medium: precondition might fail
test.adb:68:07: info: initialization of "X" proved
test.adb:68:10: info: initialization of "Y" proved
test.adb:70:07: info: precondition proved (CVC5: 1 VC)
test.adb:71:23: info: unrolling loop
test.adb:72:15: info: precondition proved (CVC5: 4 VC)
test.adb:74:07: info: precondition proved (CVC5: 1 VC)
test.adb:76:07: info: precondition proved (CVC5: 1 VC)
test.adb:77:12: info: precondition proved (CVC5: 1 VC)
test.adb:78:12: medium: precondition might fail
test.adb:79:07: info: precondition proved (Z3: 1 VC)
test.adb:80:12: medium: precondition might fail
test.adb:81:12: medium: precondition might fail
test.adb:82:07: info: precondition proved (Z3: 1 VC)
test.adb:86:07: info: initialization of "X" proved
test.adb:88:07: info: precondition proved (CVC5: 1 VC)
test.adb:89:12: info: precondition proved (CVC5: 1 VC)
test.adb:90:07: info: precondition proved (CVC5: 1 VC)
test.adb:94:07: info: initialization of "X" proved
test.adb:94:10: info: initialization of "Y" proved
test.adb:96:23: info: unrolling loop
test.adb:97:15: info: precondition proved (CVC5: 4 VC)
test.adb:100:12: info: precondition proved (CVC5: 1 VC)
test.adb:101:12: info: precondition proved (CVC5: 1 VC)
test.adb:102:07: info: precondition proved (CVC5: 1 VC)
test.adb:103:07: info: precondition proved (Z3: 1 VC)
test.adb:107:07: info: initialization of "X" proved
test.adb:107:10: info: initialization of "Y" proved
test.adb:107:13: info: initialization of "Z" proved
test.adb:109:23: info: unrolling loop
test.adb:110:15: info: precondition proved (CVC5: 4 VC)
test.adb:114:12: info: precondition proved (CVC5: 1 VC)
test.adb:115:12: info: precondition proved (CVC5: 1 VC)
test.adb:116:07: info: precondition proved (CVC5: 1 VC)
test.adb:117:12: info: precondition proved (CVC5: 1 VC)
test.adb:118:07: info: precondition proved (CVC5: 1 VC)
test.adb:122:07: info: initialization of "X" proved
test.adb:122:10: info: initialization of "Y" proved
test.adb:122:13: info: initialization of "Z" proved
test.adb:124:23: info: unrolling loop
test.adb:125:15: info: precondition proved (CVC5: 4 VC)
test.adb:127:23: info: unrolling loop
test.adb:128:15: info: precondition proved (CVC5: 4 VC)
test.adb:130:23: info: unrolling loop
test.adb:131:15: medium: precondition might fail
test.adb:133:07: medium: precondition might fail
test.adb:134:12: medium: precondition might fail
test.adb:135:07: medium: precondition might fail
test.adb:139:07: info: initialization of "X" proved
test.adb:141:07: info: precondition proved (CVC5: 1 VC)
test.adb:142:12: info: precondition proved (CVC5: 1 VC)
test.adb:143:07: info: precondition proved (CVC5: 1 VC)
test.adb:144:12: info: precondition proved (CVC5: 1 VC)
test.adb:145:07: info: precondition proved (CVC5: 1 VC)
test.adb:149:07: info: initialization of "X" proved
test.adb:149:10: info: initialization of "Y" proved
test.adb:149:13: info: initialization of "Z" proved
test.adb:151:23: info: unrolling loop
test.adb:152:15: info: precondition proved (CVC5: 3 VC)
test.adb:154:23: info: unrolling loop
test.adb:155:15: info: precondition proved (CVC5: 3 VC)
test.adb:157:23: info: unrolling loop
test.adb:158:15: medium: precondition might fail
test.adb:160:07: medium: precondition might fail
test.adb:161:12: medium: precondition might fail
test.adb:162:07: medium: precondition might fail
test.adb:163:12: info: precondition proved (Z3: 1 VC)
test.adb:164:07: info: precondition proved (Z3: 1 VC)
test.adb:168:07: info: initialization of "X" proved
test.adb:168:10: info: initialization of "Y" proved
test.adb:170:23: info: unrolling loop
test.adb:171:15: info: precondition proved (CVC5: 3 VC)
test.adb:173:23: info: unrolling loop
test.adb:174:15: info: precondition proved (CVC5: 3 VC)
test.adb:176:07: info: precondition proved (CVC5: 1 VC)
test.adb:177:12: info: precondition proved (CVC5: 1 VC)
test.adb:178:07: info: precondition proved (Z3: 1 VC)
test.adb:183:07: info: initialization of "X" proved
test.adb:183:10: info: initialization of "Y" proved
test.adb:185:07: info: precondition proved (CVC5: 1 VC)
test.adb:185:15: info: predicate check proved (Trivial: 1 VC)
test.adb:185:37: info: predicate check proved (Trivial: 1 VC)
test.adb:186:23: info: unrolling loop
test.adb:187:15: info: precondition proved (CVC5: 4 VC)
test.adb:189:07: info: precondition proved (CVC5: 1 VC)
test.adb:189:15: info: predicate check proved (Trivial: 1 VC)
test.adb:189:37: info: predicate check proved (Trivial: 1 VC)
test.adb:189:43: info: predicate check proved (Trivial: 1 VC)
test.adb:189:65: info: predicate check proved (Trivial: 1 VC)
test.adb:190:23: info: unrolling loop
test.adb:191:15: info: precondition proved (CVC5: 2 VC)
test.adb:193:07: info: precondition proved (Z3: 1 VC)
test.adb:193:15: info: predicate check proved (Trivial: 1 VC)
test.adb:193:37: info: predicate check proved (Trivial: 1 VC)
test.adb:193:43: info: predicate check proved (Trivial: 1 VC)
test.adb:193:65: info: predicate check proved (Trivial: 1 VC)
test.adb:194:23: info: unrolling loop
test.adb:195:15: medium: precondition might fail
test.adb:197:07: medium: precondition might fail
test.adb:197:15: info: predicate check proved (Trivial: 1 VC)
test.adb:197:37: info: predicate check proved (Trivial: 1 VC)
test.adb:197:43: info: predicate check proved (Trivial: 1 VC)
test.adb:197:65: info: predicate check proved (Trivial: 1 VC)
test.adb:198:23: info: unrolling loop
test.adb:199:15: info: precondition proved (Z3: 2 VC)
test.adb:201:07: medium: precondition might fail
test.adb:201:15: info: predicate check proved (Trivial: 1 VC)
test.adb:201:37: info: predicate check proved (Trivial: 1 VC)
test.adb:201:43: info: predicate check proved (Trivial: 1 VC)
test.adb:201:65: info: predicate check proved (Trivial: 1 VC)
test.adb:208:07: info: precondition proved (CVC5: 1 VC)
test.adb:208:15: info: predicate check proved (Trivial: 1 VC)
test.adb:208:28: info: predicate check proved (Trivial: 1 VC)
test.adb:213:07: info: initialization of "X" proved
test.adb:213:10: info: initialization of "Y" proved
test.adb:216:12: info: precondition proved (CVC5: 1 VC)
test.adb:217:12: info: precondition proved (CVC5: 1 VC)
test.adb:218:07: info: precondition proved (CVC5: 1 VC)
test.adb:218:15: info: predicate check proved (Trivial: 1 VC)
test.adb:218:28: info: predicate check proved (Trivial: 1 VC)
test.adb:219:07: info: precondition proved (CVC5: 1 VC)
test.adb:220:07: info: precondition proved (CVC5: 1 VC)
test.adb:225:07: info: initialization of "X" proved
test.adb:227:24: info: unrolling loop
test.adb:228:15: medium: precondition might fail
test.adb:230:12: info: precondition proved (CVC5: 1 VC)
test.adb:231:07: info: precondition proved (CVC5: 1 VC)
test.adb:231:15: info: predicate check proved (Trivial: 1 VC)
test.adb:231:28: info: predicate check proved (Trivial: 1 VC)
test.adb:232:07: info: precondition proved (CVC5: 1 VC)
test.adb:233:07: medium: precondition might fail
test.adb:237:07: info: initialization of "X" proved
test.adb:237:10: info: initialization of "Y" proved
test.adb:237:13: info: initialization of "Z" proved
test.adb:239:07: info: precondition proved (CVC5: 1 VC)
test.adb:240:23: info: unrolling loop
test.adb:241:15: info: precondition proved (CVC5: 4 VC)
test.adb:243:07: info: precondition proved (Z3: 1 VC)
test.adb:244:23: info: unrolling loop
test.adb:245:15: info: precondition proved (CVC5: 2 VC)
test.adb:247:07: info: precondition proved (Z3: 1 VC)
test.adb:249:23: info: unrolling loop
test.adb:250:15: medium: precondition might fail
test.adb:252:07: medium: precondition might fail
test.adb:253:23: info: unrolling loop
test.adb:254:15: info: precondition proved (Z3: 2 VC)
test.adb:256:07: medium: precondition might fail
test.adb:260:07: info: initialization of "X" proved
test.adb:260:10: info: initialization of "Y" proved
test.adb:260:13: info: initialization of "Z" proved
test.adb:262:07: info: precondition proved (CVC5: 1 VC)
test.adb:263:23: info: unrolling loop
test.adb:264:15: info: precondition proved (CVC5: 3 VC; Z3: 1 VC)
test.adb:266:07: info: precondition proved (CVC5: 1 VC)
test.adb:267:23: info: unrolling loop
test.adb:268:15: info: precondition proved (CVC5: 1 VC; Z3: 1 VC)
test.adb:270:07: info: precondition proved (Z3: 1 VC)
test.adb:272:23: info: unrolling loop
test.adb:273:15: medium: precondition might fail
test.adb:274:15: medium: precondition might fail
test.adb:276:07: info: precondition proved (Z3: 1 VC)
test.adb:277:23: info: unrolling loop
test.adb:278:15: info: precondition proved (Z3: 2 VC)
test.adb:280:07: info: precondition proved (Z3: 1 VC)
test.adb:284:07: info: initialization of "B" proved
test.adb:285:07: info: initialization of "X" proved
test.adb:287:24: info: unrolling loop
test.adb:288:15: medium: precondition might fail
test.adb:290:07: info: precondition proved (CVC5: 1 VC)
test.adb:291:11: info: precondition proved (CVC5: 1 VC)
test.adb:291:11: info: precondition proved (CVC5: 2 VC)
test.adb:291:11: info: precondition proved (Trivial: 1 VC)
test.adb:292:10: info: precondition proved (Z3: 1 VC)
test.adb:295:11: info: precondition proved (CVC5: 2 VC)
test.adb:295:11: info: precondition proved (Trivial: 1 VC)
test.adb:296:33: info: loop invariant initialization proved (CVC5: 1 VC; Z3: 1 VC)
test.adb:296:33: info: loop invariant preservation proved (Z3: 2 VC)
test.adb:297:30: info: precondition proved (CVC5: 1 VC)
test.adb:301:07: medium: precondition might fail
test.adb:306:18: info: elements shall be distinct for set aggregates at spark-containers-functional-sets.ads:72, in instantiation at inst.ads:4
test.adb:306:18: info: precondition proved (CVC5: 10 VC)
test.adb:308:07: info: precondition proved (CVC5: 1 VC)
test.adb:308:15: info: predicate check proved (Trivial: 1 VC)
test.adb:308:28: info: predicate check proved (Trivial: 1 VC)
test.adb:309:07: info: precondition proved (CVC5: 1 VC)
